File: leetcode-js/018-4sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[][]}
 */
var fourSum = function(nums, target) {
    'use strict';

    nums.sort((a, b) => a - b);
    const result = [];
    kSum(nums, 0, nums.length - 1, target, 4, [], result);
    return result;
};

function kSum(nums, start, end, target, k, solution, result) {
    'use strict';

    switch (k) {
    case 0:
        return;
    case 1:
        oneSum(nums, start, end, target, solution, result);
        return;
    case 2:
        twoSum(nums, start, end, target, solution, result);
        return;
    default:
        for (let i = start; i <= end - k + 1; i += 1) {
            if (i > start && nums[i] === nums[i-1]) {
                continue;
            }
            kSum(nums, i + 1, end, target - nums[i], k - 1, solution.concat(nums[i]), result);
        }
    }
}

function oneSum(nums, start, end, target, solution, result) {
    'use strict';

    for (const n of nums.slice(start, end + 1)) {
        if (n === target) {
            result.push(solution.concat(n));
        }
    }
}

function twoSum(nums, start, end, target, solution, result) {
    'use strict';

    while (start < end) {
        const sum = nums[start] + nums[end];
        if (sum === target) {
            result.push(solution.concat(nums[start], nums[end]));
            start += 1;
            end -= 1;
            while (nums[start] === nums[start-1]) {
                start += 1;
            }
            while (nums[end] === nums[end+1]) {
                end -= 1;
            }
        } else if (sum < target) {
            start += 1;
        } else {
            end -= 1;
        }
    }
}

console.log(fourSum([0, 0, 0, 0], 0));

File: leetcode-java/018-4sum.java
public class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        List<List<Integer>> res = new ArrayList<>();

        if (nums.length < 4) {
            return res;
        }

        Arrays.sort(nums);

        int max = nums[nums.length - 1];

        if (nums[0] * 4 > target || max * 4 < target) {
            return res;
        }

        for (int i = 0; i < nums.length - 3; i++) {
            int n = nums[i];

            if (i > 0 && n == nums[i - 1]) {
                continue;
            } else if (n * 4 > target) {
                break;
            } else if (n + max * 3 < target) {
                continue;
            } else if (n * 4 == target) {
                if (nums[i + 3] == n) {
                    res.add(Arrays.asList(n, n, n, n));
                }
                break;
            }

            threeSum(res, n, nums, i + 1, max, target - n);
        }

        return res;
    }

    void threeSum(List<List<Integer>> res, int a, int[] nums, int start, int max, int target) {
        if (nums[start] * 3 > target) {
            return;
        }

        for (int i = start; i < nums.length - 2; i++) {
            int n = nums[i];

            if (i > start && n == nums[i - 1]) {
                continue;
            } else if (n * 3 > target) {
                break;
            } else if (n + max * 2 < target) {
                continue;
            } else if (n * 3 == target) {
                if (nums[i + 2] == n) {
                    res.add(Arrays.asList(a, n, n, n));
                }
                break;
            }

            twoSum(res, a, n, nums, i + 1, target - n);
        }
    }

    void twoSum(List<List<Integer>> res, int a, int b, int[] nums, int start, int target) {
        int end = nums.length - 1;

        while (start < end) {
            int sum = nums[start] + nums[end];

            if (sum == target) {
                res.add(Arrays.asList(a, b, nums[start], nums[end]));
                while (start < end && nums[start] == nums[start + 1]) {
                    start++;
                }
                while (start < end && nums[end] == nums[end - 1]) {
                    end--;
                }
                start++;
                end--;
            } else if (sum < target) {
                while (start < end && nums[start] == nums[start + 1]) {
                    start++;
                }
                start++;
            } else {
                while (start < end && nums[end] == nums[end - 1]) {
                    end--;
                }
                end--;
            }
        }
    }
}
